php/libraries/SubprojectController.class.inc

```diff
@@ -0,0 +1,252 @@
+     * one object in the event that it's called multiple times.
+     *
+     * @param string $subprojectTitle The title of the subproject
+     *
+     * @return \LORIS\Subproject the Subproject object
+     */
+    public function getSubprojectByTitle($subprojectTitle) : \LORIS\Subproject
+    {
+        if (!isset(self::$_instances[$subprojectTitle])) {
+            $subprojectData = $this->database->pSelectRow(
+                "SELECT
+                    SubprojectID,
+                    title,
+                    useEDC,
+                    WindowDifference,
+                    RecruitmentTarget
+                FROM subproject
+                WHERE title = :title",
+                ['title' => $subprojectTitle]
+            );
+
+            if (empty($subprojectData)) {
+                throw new \LorisException(
+                    "No subproject with title $subprojectTitle found."
+                );
+            }
+
+            $subproject = new \LORIS\Subproject(
+                intval($subprojectData['SubprojectID']),
+                $subprojectData['title'],
+                $subprojectData['useEDC'] === '1',
+                $subprojectData['WindowDifference'],
+                isset($subprojectData['RecruitmentTarget'])
+                    ? intval($subprojectData['RecruitmentTarget']) : null
+            );
+
+            self::$_instances[$subprojectTitle] = $subproject;
+        }
+
+        return self::$_instances[$subprojectTitle];
+    }
+
+    /**
+     * Get a subproject object from the ID by mapping the ID to a title and using the
+     * singleton function
+     *
+     * @param int $subprojectID The ID of the subproject requested
+     *
+     * @return \LORIS\Subproject
+     */
+    public function getSubprojectByID(int $subprojectID) : \LORIS\Subproject
+    {
+        $subprojectTitle = $this->database->pSelectOne(
+            "SELECT
+              title
+            FROM
+              subproject
+            WHERE SubprojectID = :id
+            ",
+            ['id' => $subprojectID]
+        );
+
+        if (empty($subprojectTitle)) {
+            throw new \LorisException("No subproject with ID $subprojectID found.");
+        }
+
+        return $this->getSubprojectByTitle($subprojectTitle);
+    }
+
+    /**
+     * Create a new subproject and insert it into the database.
+     *
+     * @param \LORIS\Subproject $subproject A subproject object without the ID
+     *
+     * @return \LORIS\Subproject The created subproject with it's new ID
+     */
+    public function createSubproject(\LORIS\Subproject $subproject)
+    : \LORIS\Subproject
+    {
+        if (!empty($subproject->getID())) {
+            throw new \LorisException(
+                "An ID can not be supplied when creating a new subproject, the ID
+                will be generated by the database."
+            );
+        }
+        $subprojectExists = $this->subprojectExists($subproject->getTitle());
```

Comment:
You don't need to do that check. There will be a DatabaseException on constraint violation.
Or, if you really want that check, you should also add it in the update function.

Reply:
I didnt check if its a constraint violation (is there unique on the database side ?)
and good point for update, will change it depending on first question

Reply:
Humm, no. There is no contraint on the subproject table other than the surrogate key.
In that case, `getSubprojectByTitle` should return a Traversable of Subprojects. Or we should add a unique constraint.

Reply:
i think constraint is the way to go and we remove the soft validation I think

Reply:
@xlecours adding the constraint will send this PR to 25 directly, not adding it means we have to "expect" multiple subprojects with the same name but it could still be merged in 24...

I'm torn

Reply:
If you do it wrong and it works, it will stay until eternity ends. I am feeling philosophical :)

Reply:
no pressure... jeez!